src/Chainweb/Pact/Backend/Compaction.hs

```diff
@@ -0,0 +1,724 @@
+import GHC.Stack (HasCallStack)
+import Options.Applicative
+import System.Directory (createDirectoryIfMissing)
+import System.FilePath ((</>))
+import System.IO qualified as IO
+import System.IO (Handle)
+
+import Chainweb.BlockHeight (BlockHeight)
+import Chainweb.Logger (setComponent)
+import Chainweb.Utils (sshow, HasTextRepresentation, fromText, toText, int)
+import Chainweb.Version (ChainId, ChainwebVersion(..), ChainwebVersionName, unsafeChainId, chainIdToText)
+import Chainweb.Version.Mainnet (mainnet)
+import Chainweb.Version.Registry (lookupVersionByName)
+import Chainweb.Version.Utils (chainIdsAt)
+import Chainweb.Pact.Backend.Types (SQLiteEnv(..))
+import Chainweb.Pact.Backend.Utils (fromUtf8, withSqliteDb)
+import Chainweb.Utils (encodeB64Text)
+
+import System.Logger
+import System.Logger.Backend.ColorOption (useColor)
+import Data.LogMessage
+
+import Pact.Types.Persistence (TxId(..))
+import Pact.Types.SQLite (SType(..), RType(..))
+import Pact.Types.SQLite qualified as Pact
+
+newtype TableName = TableName { getTableName :: Utf8 }
+  deriving stock (Show)
+  deriving newtype (Eq, IsString)
+
+data CompactException
+  = CompactExceptionInternal !Text
+  | CompactExceptionDb !SomeException
+  | CompactExceptionInvalidBlockHeight !BlockHeight
+  | CompactExceptionTableVerificationFailure !TableName
+  | CompactExceptionNoLatestBlockHeight
+  deriving stock (Show)
+  deriving anyclass (Exception)
+
+data CompactFlag
+  = KeepCompactTables
+    -- ^ Keep compaction tables post-compaction for inspection.
+  | NoVacuum
+    -- ^ Don't VACUUM database
+  | NoDropNewTables
+    -- ^ Don't drop new tables created after the compaction height.
+  | NoGrandHash
+    -- ^ Don't compute the grand hash.
+  deriving stock (Eq,Show,Read,Enum,Bounded)
+
+internalError :: MonadThrow m => Text -> m a
+internalError = throwM . CompactExceptionInternal
+
+data CompactEnv = CompactEnv
+  { _ceDb :: !Database
+  , _ceLogger :: !(Logger SomeLogMessage)
+  , _ceFlags :: ![CompactFlag]
+  }
+makeLenses ''CompactEnv
+
+withDefaultLogger :: LogLevel -> (Logger SomeLogMessage -> IO a) -> IO a
+withDefaultLogger ll f = withHandleBackend_ logText defaultHandleBackendConfig $ \b ->
+    withLogger defaultLoggerConfig b $ \l -> f (set setLoggerLevel ll l)
+
+withPerChainFileLogger :: FilePath -> ChainId -> LogLevel -> (Logger SomeLogMessage -> IO a) -> IO a
+withPerChainFileLogger logDir chainId ll f = do
+  createDirectoryIfMissing False {- don't create parents -} logDir
+  let logFile = logDir </> ("chain-" <> cid <> ".log")
+  !_ <- writeFile logFile ""
+  let handleConfig = defaultHandleBackendConfig
+        { _handleBackendConfigHandle = FileHandle logFile
+        }
+  withHandleBackend_' logText handleConfig $ \h b -> do
+
+    done <- newMVar False
+    void $ forkIO $ fix $ \go -> do
+      doneYet <- readMVar done
+      when (not doneYet) $ do
+        threadDelay 5_000_000
+        IO.hFlush h
+        go
+      IO.hFlush h
```

Comment:
Just noticed that this will induce a very small space leak, because it's not a tail call. Perhaps put `hFlush` at the start instead?